### PR TITLE
Refactor workflow execution context & mutable state

### PR DESCRIPTION
**Pull request by wxing1292**
* Remove redundant append first event batch APIs
* Remove mutableStateSessionUpdates in favor of WorkflowSnapshot & WorkflowMutation
* Move force fail decision for failover / buffered events size limit into CloseTransactionAsMutation & CloseTransactionAsSnapshot
* Mutable state will now keep track of all transfer / timer tasks by AddTransferTasks & AddTimerTasks
* Remove the requirement for transactionID from shard when create / update workflow
* Remove special handling of LimitExceededError since buffered events size limit is handled before close transaction

### CODE REVIEW

File: service/history/mutableStateBuilder.go [hunk 1]
```diff
@@ -3684,31 +3671,39 @@ func (e *mutableStateBuilder) prepareEventsAndReplicationTasks() ([]*persistence
...
-		convertSyncActivityInfos(
-			e.pendingActivityInfoIDs,
-			e.syncActivityTasks,
-		)...)
+		e.syncActivityToReplicationTask(closeAsActive)...,
+	)
+
+	if !closeAsActive && len(e.insertReplicationTasks) > 0 {
+		return nil, &workflow.InternalServiceError{
+			Message: "should no generate replication task when close transaction as passive",
```
Comment (venkat1109) [inline service/history/mutableStateBuilder.go]: should not*

File: service/history/mutableStateBuilder.go [hunk 2]
```diff
@@ -3777,6 +3792,122 @@ func (e *mutableStateBuilder) validateNoEventsAfterWorkflowFinish(
 	}
 }
 
+func (e *mutableStateBuilder) closeTransactionCheckActiveness(
```
Comment (venkat1109) [inline service/history/mutableStateBuilder.go]: consider renaming to closeTransactionIfActive()

File: service/history/workflowExecutionContext.go [hunk 3]
```diff
@@ -151,11 +134,10 @@ type (
 		metricsClient     metrics.Client
 		timeSource        clock.TimeSource
 
-		locker                locks.Mutex
-		msBuilder             mutableState
-		stats                 *persistence.ExecutionStats
-		updateCondition       int64
-		createReplicationTask bool
+		locker          locks.Mutex
```
Comment (venkat1109) [inline service/history/workflowExecutionContext.go]: nit: mutex is a well understood name for this

File: service/history/conflictResolver.go [hunk 4]
```diff
@@ -119,6 +119,9 @@ func (r *conflictResolverImpl) reset(
 				r.shard.GetEventsCache(),
 				r.logger,
 				firstEvent.GetVersion(),
+				// if can see replication task, meaning that domain is
+				// global domain with > 1 target clusters
+				true,
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/conflictResolver.go]: this is on the replication receiver side, meaning that if logic is being executed, it is guaranteed that the mutable state to be generated can do replication

**Pull request merged**